File: benchmarks/primes.lua
local UPPER_BOUND = 5000000
local PREFIX = 32338

local Node = {}
function Node:new()
    local obj = {}
    setmetatable(obj, self)
    self.__index = self
    obj.children = {}
    obj.terminal = false
    return obj
end

local Sieve = {}
function Sieve:new(limit)
    local obj = {}
    setmetatable(obj, self)
    self.__index = self
    obj.limit = limit
    obj.prime = {}
    for i = 0, limit do
        obj.prime[i] = false
    end
    return obj
end

function Sieve:to_list()
    local result = {2, 3}
    for p = 5, self.limit do
        if self.prime[p] then
            table.insert(result, p)
        end
    end
    return result
end

function Sieve:omit_squares()
    local r = 5
    while r * r < self.limit do
        if self.prime[r] then
            local i = r * r
            while i < self.limit do
                self.prime[i] = false
                i = i + r * r
            end
        end
        r = r + 1
    end
    return self
end

function Sieve:step1(x, y)
    local n = (4 * x * x) + (y * y)
    if n <= self.limit and (n % 12 == 1 or n % 12 == 5) then
        self.prime[n] = not self.prime[n]
    end
end

function Sieve:step2(x, y)
    local n = (3 * x * x) + (y * y)
    if n <= self.limit and n % 12 == 7 then
        self.prime[n] = not self.prime[n]
    end
end

function Sieve:step3(x, y)
    local n = (3 * x * x) - (y * y)
    if x > y and n <= self.limit and n % 12 == 11 then
        self.prime[n] = not self.prime[n]
    end
end

function Sieve:loop_y(x)
    local y = 1
    while y * y < self.limit do
        self:step1(x, y)
        self:step2(x, y)
        self:step3(x, y)
        y = y + 1
    end
end

function Sieve:loop_x()
    local x = 1
    while x * x < self.limit do
        self:loop_y(x)
        x = x + 1
    end
end

function Sieve:calc()
    self:loop_x()
    return self:omit_squares()
end

local function generate_trie(l)
    local root = Node:new()
    for _, el in ipairs(l) do
        local head = root
        -- attempt to call a nil value (method 'split')
        -- how to fix? use string.split
        el = tostring(el)
        for i=1, #el do
            local ch = el:sub(i, i)
            if not head.children[ch] then
                head.children[ch] = Node:new()
            end
            head = head.children[ch]
        end
        head.terminal = true
    end
    return root
end

local function find(upper_bound, prefix_)
    local primes = Sieve:new(upper_bound):calc()
    local str_prefix = tostring(prefix_)
    local head = generate_trie(primes:to_list())
    for i=1, #str_prefix do
        local ch = str_prefix:sub(i, i)
        head = head.children[ch]
        if head == nil then
            return nil
        end
    end

    local queue, result = {{head, str_prefix}}, {}
    while #queue > 0 do
        local tuple = table.remove(queue)
        local top, prefix = tuple[1], tuple[2]
        if top.terminal then
            table.insert(result, tonumber(prefix))
        end
        for ch, v in pairs(top.children) do
            table.insert(queue, 1, {v, prefix .. ch})
        end
    end

    table.sort(result)
    return result
end

local function verify()
    local left = {2, 23, 29}
    local right = find(100, 2)
    if #left ~= #right then
        print("length not equal")
        os.exit(1)
    end
    for i, v in ipairs(left) do
        if v ~= right[i] then
            print(string.format("%s != %s", v, right[i]))
            os.exit(1)
        end
    end
end

verify()
local results = find(UPPER_BOUND, PREFIX)
local expected = {323381, 323383, 3233803, 3233809, 3233851, 3233863, 3233873, 3233887, 3233897}

for i, v in ipairs(results) do
    if v ~= expected[i] then
        print(string.format("%s != %s", v, expected[i]))
        os.exit(1)
    end
end

File: benchmarks/primes.cpp
#include <vector>
#include <string>
#include <queue>
#include <iostream>
#include <algorithm>
#include <unordered_map>

typedef long long LL;

struct Node{
    std::unordered_map<int, Node*> children;
    bool terminal;

    Node(){
        terminal = false;
    }
};

struct Sieve{
    LL limit;
    std::vector<bool> prime;

    Sieve(LL limit){
        this->limit = limit;
        prime = std::vector<bool>(limit + 1, false);
    }

    std::vector<LL> to_list(){
        std::vector<LL> result;
        result.push_back(2);
        result.push_back(3);
        for(LL p = 5; p <= limit; p++){
            if(prime[p]){
                result.push_back(p);
            }
        }
        return result;
    }

    void omit_squares(){
        LL r = 5;
        while(r * r < limit){
            if(prime[r]){
                LL i = r * r;
                while(i < limit){
                    prime[i] = false;
                    i = i + r * r;
                }
            }
            r += 1;
        }
    }

    void step1(LL x, LL y){
        LL n = (4 * x * x) + (y * y);
        if(n <= limit && (n % 12 == 1 || n % 12 == 5)){
            prime[n] = !prime[n];
        }
    }

    void step2(LL x, LL y){
        LL n = (3 * x * x) + (y * y);
        if(n <= limit && n % 12 == 7){
            prime[n] = !prime[n];
        }
    }

    void step3(LL x, LL y){
        LL n = (3 * x * x) - (y * y);
        if(x > y && n <= limit && n % 12 == 11){
            prime[n] = !prime[n];
        }
    }

    void loop_y(LL x){
        LL y = 1;
        while(y * y < limit){
            step1(x, y);
            step2(x, y);
            step3(x, y);
            y += 1;
        }
    }

    void loop_x(){
        LL x = 1;
        while(x * x < limit){
            loop_y(x);
            x += 1;
        }
    }

    void calc(){
        loop_x();
        omit_squares();
    }
};

Node *generate_trie(std::vector<LL> l){
    Node *root = new Node();
    for(LL el : l){
        Node *head = root;
        std::string s = std::to_string(el);
        for(char ch : s){
            if(head->children.find(ch) == head->children.end()){
                head->children[ch] = new Node();
            }
            head = head->children[ch];
        }
        head->terminal = true;
    }
    return root;
}

std::vector<LL> find(LL upper_bound, LL prefix){
    Sieve *sieve = new Sieve(upper_bound);
    sieve->calc();
    std::string str_prefix = std::to_string(prefix);
    Node *head = generate_trie(sieve->to_list());
    for(char ch : str_prefix){
        if(head->children.find(ch) == head->children.end()){
            return std::vector<LL>();
        }
        head = head->children[ch];
    }

    std::queue<std::pair<Node*, std::string>> q;
    std::vector<LL> result;
    q.push(std::make_pair(head, str_prefix));
    while(!q.empty()){
        std::pair<Node*, std::string> top = q.front();
        q.pop();
        if(top.first->terminal){
            result.push_back(std::stoll(top.second));
        }
        for(std::pair<char, Node*> p : top.first->children){
            q.push(std::make_pair(p.second, top.second + p.first));
        }
    }

    std::sort(result.begin(), result.end());
    return result;
}

void verify(){
    std::vector<LL> left = {2, 23, 29};
    std::vector<LL> right = find(100, 2);
    if(left != right){
        std::cout << "left != right" << std::endl;
        exit(1);
    }
}


int main(){
    const LL UPPER_BOUND = 5000000;
    const LL PREFIX = 32338;

    verify();
    std::vector<LL> results = find(UPPER_BOUND, PREFIX);
    std::vector<LL> expected = {323381, 323383, 3233803, 3233809, 3233851, 3233863, 3233873, 3233887, 3233897};
    if(results != expected){
        std::cout << "results != expected" << std::endl;
        exit(1);
    }
    return 0;
}